File: MtxVec.Delphi/ExprToolTipUnit.pas
unit ExprToolTipUnit;

interface

uses
  Windows, Messages, SysUtils, Variants, Graphics,
  Controls, Forms, Dialogs, ComCtrls, Grids, MtxVec, MtxVecInt,
  MtxGrid, MtxVecBase, DateUtils, Classes;

type
  TExprToolTipForm = class(TForm)
    MtxVecGrid: TMtxVecGrid;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
  private
    bName: string;
    mark: TDateTime;
    FData: TMtx;
    procedure SetData(const Value: TMtx);
  protected
    procedure WMNCHitTest(var Message: TWMNCHitTest); message WM_NCHITTEST;
    { Private declarations }
  public
    property Data: TMtx read FData write SetData;
    procedure UpdateGrid(const Src: TMtxVecBase; const aName: string);
    { Public declarations }
  end;

var
  ExprToolTipForm: TExprToolTipForm;

implementation

{$R *.dfm}

procedure TExprToolTipForm.FormCreate(Sender: TObject);
begin
//    FData := TMtx.Create;
//    MtxVecGrid.Datasource := FData;
end;

procedure TExprToolTipForm.FormDestroy(Sender: TObject);
begin
//    FData.Free;
end;

procedure TExprToolTipForm.SetData(const Value: TMtx);
begin
//  FData := Value;
end;

procedure TExprToolTipForm.UpdateGrid(const Src: TMtxVecBase; const aName: string);
begin
    if aName = bName then
    begin
        if (Now - Mark) > (0.25*OneSecond) then
        begin
          {  case Src.MtxVecType of
            mvTVec: begin
                    FData.Size(TVec(Src).Length, 1, TVec(Src).Complex);
                    FData.Copy(TVec(Src));
                    end;
            mvTMtx: FData.Copy(TMtx(Src));
            mvTVecInt: begin
                       FData.Size(TVecInt(Src).Length, 1);
                       FData.Copy(TVecInt(Src));
                       end;
            mvTMtxInt: FData.Copy(TMtxInt(Src))
            end;  }
            MtxVecGrid.Datasource := Src;
            Show;
            Mark := Now;
        end;
    end else
    begin
        Mark := Now;
        bName := aName;
    end;
end;

procedure TExprToolTipForm.WMNCHitTest(var Message: TWMNCHitTest);
const EDGEDETECT = 7; // adjust
var deltaRect: TRect;
begin
  inherited;
  if BorderStyle = bsNone then
    with Message, deltaRect do
    begin
      Left := XPos - BoundsRect.Left;
      Right := BoundsRect.Right - XPos;
      Top := YPos - BoundsRect.Top;
      Bottom := BoundsRect.Bottom - YPos;
      if (Top < EDGEDETECT) and (Left < EDGEDETECT) then
        Result := HTTOPLEFT
      else if (Top < EDGEDETECT) and (Right < EDGEDETECT) then
        Result := HTTOPRIGHT
      else if (Bottom < EDGEDETECT) and (Left < EDGEDETECT) then
        Result := HTBOTTOMLEFT
      else if (Bottom < EDGEDETECT) and (Right < EDGEDETECT) then
        Result := HTBOTTOMRIGHT
      else if (Top < EDGEDETECT) then
        Result := HTTOP
      else if (Left < EDGEDETECT) then
        Result := HTLEFT
      else if (Bottom < EDGEDETECT) then
        Result := HTBOTTOM
      else if (Right < EDGEDETECT) then
        Result := HTRIGHT
    end;
end;

end.

File: MtxVec.CPP/ExprToolTipUnit.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "ExprToolTipUnit.h"
#include <DateUtils.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TExprToolTipForm *ExprToolTipForm;
//---------------------------------------------------------------------------
__fastcall TExprToolTipForm::TExprToolTipForm(TComponent* Owner)
  : TForm(Owner), FData(NULL)
{
}
//---------------------------------------------------------------------------
void __fastcall TExprToolTipForm::FormCreate(TObject *Sender)
{
//  FData = new TMtx();
//  MtxVecGrid->Datasource = FData;
}
//---------------------------------------------------------------------------
void __fastcall TExprToolTipForm::FormDestroy(TObject *Sender)
{
//  delete FData;
}
//---------------------------------------------------------------------------
void __fastcall TExprToolTipForm::SetData(TMtx * const Value)
{
//  FData = Value;
}
//---------------------------------------------------------------------------
void __fastcall TExprToolTipForm::UpdateGrid(TMtxVecBase * const Src, const AnsiString &aName)
{
  if (aName == bName)
  {
    if ((double)(Now() - mark) > (0.25 * (double)OneSecond))
    {
      MtxVecGrid->Datasource = Src;
      Show();
      mark = Now();
    }
  }
  else
  {
    mark = Now();
    bName = aName;
  }
}
//---------------------------------------------------------------------------
void __fastcall TExprToolTipForm::WMNCHitTest(TWMNCHitTest &Message)
{
  const int EDGEDETECT = 7;
  TForm::Dispatch(&Message);
  if (BorderStyle == bsNone)
  {
    TRect deltaRect;
    deltaRect.Left = Message.XPos - BoundsRect.Left;
    deltaRect.Right = BoundsRect.Right - Message.XPos;
    deltaRect.Top = Message.YPos - BoundsRect.Top;
    deltaRect.Bottom = BoundsRect.Bottom - Message.YPos;

    if (deltaRect.Top < EDGEDETECT && deltaRect.Left < EDGEDETECT)
      Message.Result = HTTOPLEFT;
    else if (deltaRect.Top < EDGEDETECT && deltaRect.Right < EDGEDETECT)
      Message.Result = HTTOPRIGHT;
    else if (deltaRect.Bottom < EDGEDETECT && deltaRect.Left < EDGEDETECT)
      Message.Result = HTBOTTOMLEFT;
    else if (deltaRect.Bottom < EDGEDETECT && deltaRect.Right < EDGEDETECT)
      Message.Result = HTBOTTOMRIGHT;
    else if (deltaRect.Top < EDGEDETECT)
      Message.Result = HTTOP;
    else if (deltaRect.Left < EDGEDETECT)
      Message.Result = HTLEFT;
    else if (deltaRect.Bottom < EDGEDETECT)
      Message.Result = HTBOTTOM;
    else if (deltaRect.Right < EDGEDETECT)
      Message.Result = HTRIGHT;
  }
}
//---------------------------------------------------------------------------
